File: android/src/main/java/com/reactnativemembrane/VideoPreviewView.kt
package com.reactnativemembrane

import android.content.Context
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.views.ExpoView
import org.membraneframework.rtc.ui.VideoTextureViewRenderer
import org.membraneframework.rtc.media.LocalVideoTrack
import org.webrtc.PeerConnectionFactory
import org.membraneframework.rtc.media.VideoParameters
import org.webrtc.EglBase
import org.webrtc.RendererCommon

class VideoPreviewView(context: Context, appContext: AppContext) : ExpoView(context, appContext) {
    private var localVideoTrack: LocalVideoTrack? = null
    private var eglBase: EglBase? = null
    private var peerConnectionFactory: PeerConnectionFactory? = null
    private var isInitialized: Boolean = false

    private var captureDeviceId: String? = null

    private val videoView = VideoTextureViewRenderer(context).also {
        addView(it)
    }

    private fun initialize() {
        if(isInitialized) return
        isInitialized = true
        PeerConnectionFactory.initialize(PeerConnectionFactory.InitializationOptions.builder(context).createInitializationOptions())
        val eglBase = EglBase.create()
        this.eglBase = eglBase
        val peerConnectionFactory = PeerConnectionFactory.builder().createPeerConnectionFactory()
        this.peerConnectionFactory = peerConnectionFactory

        localVideoTrack = LocalVideoTrack.create(context, peerConnectionFactory, eglBase, VideoParameters.presetFHD169).also {
            it.start()
        }
        videoView.init(eglBase.eglBaseContext, null)
        localVideoTrack?.addRenderer(videoView)
    }

    fun switchCamera(captureDeviceId: String) {
        this.captureDeviceId = captureDeviceId
        localVideoTrack?.switchCamera(captureDeviceId)
    }

    fun setVideoLayout(videoLayout: String) {
        val scalingType = when (videoLayout) {
            "FILL" -> RendererCommon.ScalingType.SCALE_ASPECT_FILL
            "FIT" -> RendererCommon.ScalingType.SCALE_ASPECT_FIT
            else -> RendererCommon.ScalingType.SCALE_ASPECT_FILL
        }
        videoView.setScalingType(scalingType)
        videoView.setEnableHardwareScaler(true)
    }

    fun setMirrorVideo(mirrorVideo: Boolean) {
        videoView.setMirror(mirrorVideo)
    }

    private fun dispose() {
        localVideoTrack?.removeRenderer(videoView)
        localVideoTrack?.stop()
        localVideoTrack?.rtcTrack()?.dispose()
        videoView.release()
        peerConnectionFactory?.dispose()
        eglBase?.release()
        isInitialized = false
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        dispose()
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        initialize()
    }
}

File: ios/VideoPreviewView.swift
import ExpoModulesCore
import MembraneRTC

class VideoPreviewView: ExpoView {
  var videoView: VideoView? = nil
  private var localVideoTrack: LocalCameraVideoTrack? = nil
  
  required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    videoView = VideoView()
    videoView?.autoresizingMask = [.flexibleWidth, .flexibleHeight]
    videoView?.clipsToBounds = true
    addSubview(videoView!)
  }
  
  override func willMove(toWindow newWindow: UIWindow?) {
    if(newWindow == nil) {
      localVideoTrack?.stop()
    } else {
      localVideoTrack = LocalVideoTrack.create(videoParameters: VideoParameters.presetFHD169) as? LocalCameraVideoTrack
      localVideoTrack?.start()
      videoView?.track = localVideoTrack
    }
  }

  var videoLayout: String = "FILL" {
    didSet {
      switch(videoLayout) {
      case "FIT":
        self.videoView?.layout = .fit
      case "FILL":
        self.videoView?.layout = .fill
      default:
        self.videoView?.layout = .fill
      }
    }
  }
  
  var mirrorVideo: Bool = false {
    didSet {
      self.videoView?.mirror = mirrorVideo
    }
  }
  
  var captureDeviceId: String? = nil {
    didSet {
      if let captureDeviceId = captureDeviceId {
        localVideoTrack?.switchCamera(deviceId: captureDeviceId)
      }
    }
  }
}
